### PR TITLE
commands: remove EnableStdin support for StringArg [v2]

### PR DESCRIPTION
**Pull request by thomas-gardner**
With verbose flag:
- remove EnableStdin() flags on all StringArg,

- remove all unneeded parsing code for StringArg, and print an
informative message if `ipfs` begins reading from a CharDevice,

- remove broken go tests for EnableStdin cli parsing, and add some
trivial test cases for reading FileArg from stdin,

- add a panic to prevent EnableStdin from being set on
StringArg in the future.

Resolves: #2877, #2870, #2868
License: MIT
Signed-off-by: Thomas Gardner <tmg@fastmail.com>

### CODE REVIEW

File: commands/cli/parse.go [hunk 1]
```diff
@@ -325,6 +301,10 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 				var file files.File
 				var err error
 				if fpath == "-" {
+					stdInfo, _ := stdin.Stat()
```
Comment (Kubuxu) [inline commands/cli/parse.go]: I know that this error shouldn't happen in practice but we should check for it either way.
Log it under log.Error with some "This shouldn't happen" message.

File: commands/cli/parse.go [hunk 2]
```diff
@@ -337,6 +317,10 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 			} else {
 				if stdin != nil && argDef.SupportsStdin &&
 					argDef.Required && !fillingVariadic {
+					stdInfo, _ := stdin.Stat()
```
Comment (Kubuxu) [inline commands/cli/parse.go]: Same here, maybe extract it to function "isTerminal" or "printReadInfo" so it doesn't repeat.

**Comment by Kubuxu** [comment]
@thomas-gardner great work on that, I left a comment and thing left to do is changing sharness tests to accommodate this change. Also I am writing down why we decided to make this change and what we are changing.

**Comment by Kubuxu** [comment]
In this PR we are removing support of direct stdin handling for non-file type arguments in go-ipfs. There are two main reasons why we decided for it:
* making non-file type arguments being piped into `ipfs` process behave as expected is very difficult task. It is connected with fact that on CLI side there is very limited logic running and this logic can't decided if some argument will be used or not. This lead to introduction of regression that caused `--help` to hang (https://github.com/ipfs/go-ipfs/issues/2868). Some of you will say: "It was working before", my response is: "No it wasn't". We were cheating by treating console differently from other types of inputs which became a problem with docker and embedded nodes.
* behaviour like that is unprecedented before in any other command. It can be replicated with much more flexibility using the `xargs` command, is explicit and familiar to users and doesn't expose us to implementation edge cases.

Those are two main reasons for removal of direct stdin handling for non-file type arguments in `ipfs` command.

**Comment by Kubuxu** [comment]
As I am really co-creator of this issue/PR @whyrusleeping you would have to review it.

File: commands/cli/parse_test.go [hunk 3]
```diff
@@ -259,53 +247,17 @@ func TestArgumentParsing(t *testing.T) {
...
 		}
-		defer os.Remove(fstdin.Name())
 
 		if _, err := io.WriteString(fstdin, content); err != nil {
 			t.Fatal(err)
 		}
 		return fstdin
 	}
-
-	test([]string{"stdinenabled", "value1", "value2"}, nil, []string{"value1", "value2"})
```
Comment (whyrusleeping) [inline commands/cli/parse_test.go]: Are you sure that all these tests need to be removed?

File: commands/cli/parse.go [hunk 4]
```diff
@@ -12,9 +11,12 @@ import (
 
 	cmds "github.com/ipfs/go-ipfs/commands"
 	files "github.com/ipfs/go-ipfs/commands/files"
+	logging "gx/ipfs/QmYtB7Qge8cJpXc4irsEp8zRqfnZMBeB7aTrMEkPk67DRv/go-log"
```
Comment (Kubuxu) [inline commands/cli/parse.go]: This is wrong go-log version. You either used old version or need to rebase on top of master.

File: commands/cli/parse_test.go (hunk 3 above)
Comment (whyrusleeping) [inline commands/cli/parse_test.go]: We definitely want to have solid coverage over this. I'm not in favor of mass removing tests like this

Comment (whyrusleeping) [inline commands/cli/parse_test.go]: (i noticed you added a bunch in now)

**Comment by Kubuxu** [comment]
LGTM

File: test/sharness/t0040-add-and-cat.sh [hunk 5]
```diff
@@ -223,8 +223,8 @@ test_expect_success "ipfs cat output looks good" '
 	test_cmp expected actual
 '
 
-test_expect_success "ipfs cat accept hash from stdin" '
-	echo "$HASH" | ipfs cat >actual
+test_expect_success "ipfs cat accept hash from built input" '
+	echo "$HASH" | xargs ipfs cat >actual
```
Comment (jbenet) [inline test/sharness/t0040-add-and-cat.sh]: I thought this had been reverted, @whyrusleeping @Kubuxu -- the tests on master show it's still here.

Comment (jbenet) [inline test/sharness/t0040-add-and-cat.sh]: I want this test file have 0 (or minimal) changes in rc4.

**Pull request merged**